packages/credentials/src/status-list-credential.ts

```diff
@@ -0,0 +1,238 @@
+    credentials: VerifiableCredential[]
+  ): string[] {
+    const duplicateSet = new Set<string>();
+    for (const vc of credentials) {
+      if (!vc.vcDataModel.credentialStatus) {
+        throw new Error('no credential status found in credential');
+      }
+
+      const statusListEntry: StatusList2021Entry = vc.vcDataModel.credentialStatus as StatusList2021Entry;
+
+      if (statusListEntry.statusPurpose !== statusPurpose) {
+        throw new Error('status purpose mismatch');
+      }
+
+      if (duplicateSet.has(statusListEntry.statusListIndex)) {
+        throw new Error(`duplicate entry found with index: ${statusListEntry.statusListIndex}`);
+      }
+
+      if(parseInt(statusListEntry.statusListIndex) < 0) {
+        throw new Error('status list index cannot be negative');
+      }
+
+      if(parseInt(statusListEntry.statusListIndex) >= BITSTRING_SIZE) {
+        throw new Error('status list index is larger than the bitset size');
+      }
+
+      duplicateSet.add(statusListEntry.statusListIndex);
+    }
+
+    return Array.from(duplicateSet);
+  }
+
+  /**
+   * Generates a compressed bitstring from an array of statusListIndex values.
+   *
+   * @param statusListIndexes - An array of statusListIndex values.
+   * @returns {string} The compressed bitstring as a base64-encoded string.
+   */
+  private static bitstringGeneration(statusListIndexes: string[]): string {
+    // Initialize a Buffer with 16KB filled with zeros
+    const bitstring = new Uint8Array(BITSTRING_SIZE / 8);
+
+    // Set bits for revoked credentials
+    statusListIndexes.forEach(index => {
+      const statusListIndex = parseInt(index);
+      const byteIndex = Math.floor(statusListIndex / 8);
+      const bitIndex = statusListIndex % 8;
+
+      bitstring[byteIndex] = bitstring[byteIndex] | (1 << (7 - bitIndex)); // Set bit to 1
+    });
+
+    // Compress the bitstring with GZIP using pako
+    const compressed = pako.gzip(bitstring);
+
+    // Return the base64-encoded string
+    const base64EncodedString = Convert.uint8Array(compressed).toBase64Url();
+
+    return base64EncodedString;
+  }
+
+  /**
+   * Expands a compressed bitstring into an array of 0s and 1s.
+   *
+   * @param compressedBitstring - The compressed bitstring as a base64-encoded string.
+   * @returns {number[]} An array of 0s and 1s representing the bitstring.
+   */
+  private static bitstringExpansion(compressedBitstring: string): number[] {
+    // Base64-decode the compressed bitstring
+    const compressedData = Convert.base64Url(compressedBitstring).toUint8Array();
+
+    // Decompress the data using pako
+    const decompressedData = pako.inflate(compressedData);
+
+    // Convert the decompressed data into an array of "0" or "1" strings
+    const bitstringArray: number[] = [];
+    decompressedData.forEach(byte => {
+      for (let i = 7; i >= 0; i--) {
+        const bit = (byte >> i) & 1;
+        bitstringArray.push(bit);
+      }
+    });
```

Comment:
This is pretty inefficient: will iterate ~160,000 times minimum (assuming 16KB size) when can be done with an O(1) look up.

Consider an alternative imple by changing the method from `bitstringExpansion()` to something like `getBit(bitArray, bitIndex): boolean`, something like:

``` ts
      // Find the byte index, and bit index within the byte.
    const byteIndex = Math.floor(bitIndex / 8);
    const bitIndexWithinByte = bitIndex % 8;

    const byte = decompressedData[byteIndex];

    const bitInteger = (byte >> bitIndexWithinByte) & 1; // Remove all preceding bits using bit shift, then remove trailing bits using &.
    return (bitInteger === 1);
```

Reply:
genius! yes much better